# Triclosan tolerance is driven by a conserved mechanism in diverse Pseudomonas species

For help or any data requests, please contact Alex McFarland.
email: alexandermcfarland2022@u.northwestern.edu


### The repository for this project four folders with the data and R script used to analyze the data:

1. ani_anlysis
2. phylogneetic_analysis
3. quality_assessments
4. susceptibility analysis

In order to run the analyses yourself, please do the following:

Download the data and script from the folder.

Unzip the data

Change the main_path variable to the folder containing the downloeaded data. 

Run the analyses. 


R was run version 4.0.2 on a x86_64-apple-darwin17.0 (64-bit) platform. These scripts were tested to work using R ersion 4.0.2 on MacOS (x86_64-apple-darwin17.0) (64-bit).


Please contact me if there are issues downlaoding certain libraries.


### Additional info

The concatenated 34 gene marker tree is located in the phylogenetic_analysis with the filename __RAxML_bipartitions.conctat34mlsa__.


The data used to generate mauve whole genome anlignments are in the __mauve_alignments__ folder.


All raw figures (not arranged/beautified in illustrator and used in the final version of the paper) are located in the __figures__ folder.

FabV sequences detected in the 160 type strains, 30 study isolates, or 7,163 *Pseudomonas* genomes are located in the __fabv_seqs_phylogeny__ folder.


The python code used to work with genomes, sequence data, islandviewer and iceberg submission, are loated in the __python_scripts__ folder. 





